bot/exts/moderation/silence.py

```diff
@@ -92,88 +112,222 @@ async def _async_init(self) -> None:
+
+        else:
+            role = self._verified_voice_role
+            overwrite = channel.overwrites_for(role)
+            prev_overwrites = dict(speak=overwrite.speak)
+            if kick:
+                prev_overwrites.update(connect=overwrite.connect)
 
+        # Stop if channel was already silenced
         if channel.id in self.scheduler or all(val is False for val in prev_overwrites.values()):
             return False
 
-        overwrite.update(send_messages=False, add_reactions=False)
-        await channel.set_permissions(self._everyone_role, overwrite=overwrite)
+        # Set new permissions, store
+        overwrite.update(**dict.fromkeys(prev_overwrites, False))
+        await channel.set_permissions(role, overwrite=overwrite)
         await self.previous_overwrites.set(channel.id, json.dumps(prev_overwrites))
 
         return True
 
-    async def _schedule_unsilence(self, ctx: Context, duration: Optional[int]) -> None:
+    @staticmethod
+    async def _get_afk_channel(guild: Guild) -> VoiceChannel:
+        """Get a guild's AFK channel, or create one if it does not exist."""
+        afk_channel = guild.afk_channel
+
+        if afk_channel is None:
+            overwrites = {
+                guild.default_role: PermissionOverwrite(speak=False, connect=False, view_channel=False)
+            }
+            afk_channel = await guild.create_voice_channel("mute-temp", overwrites=overwrites)
+            log.info(f"Failed to get afk-channel, created temporary channel #{afk_channel} ({afk_channel.id})")
+
+        return afk_channel
+
+    async def _kick_voice_members(self, channel: VoiceChannel) -> None:
+        """Remove all non-staff members from a voice channel."""
+        log.debug(f"Removing all non staff members from #{channel.name} ({channel.id}).")
+
+        for member in channel.members:
+            if self._helper_role not in member.roles:
+                try:
+                    await member.move_to(None, reason="Kicking member from voice channel.")
+                    log.debug(f"Kicked {member.name} from voice channel.")
+                except Exception as e:
+                    log.debug(f"Failed to move {member.name}. Reason: {e}")
+                    continue
+
+        log.debug("Removed all members.")
+
+    async def _force_voice_sync(self, channel: VoiceChannel) -> None:
+        """
+        Move all non-staff members from `channel` to a temporary channel and back to force toggle role mute.
+
+        Permission modification has to happen before this function.
+        """
+        # Obtain temporary channel
+        delete_channel = channel.guild.afk_channel is None
+        afk_channel = await self._get_afk_channel(channel.guild)
+
+        try:
+            # Move all members to temporary channel and back
+            for member in channel.members:
+                # Skip staff
+                if self._helper_role in member.roles:
```

Comment:
This should affect anyone who doesn't have explicit connect and speak permissions in the channel, as in text. For simplicity, lets assume it's anyone who isn't mod+. This should use the has_any_roles check with the moderation roles, as in other parts of the codebase.

Reply:
Using the `has_any_role_check` without passing a fake context didn't seem feasible, so I opted to use the `any(role.id in Moderation_Roles)` structure seen elsewhere in the codebase.

Come to think of it, that should become it's own helper, but that's out of scope for this PR.